openvpn/build/build.py

```diff
@@ -1,139 +1,205 @@
 OPENVPN_GUI_REPO_DIRNAME = 'openvpn-gui'
-OPENVPN_LINUX_PREFIX = '/usr/local/openvpn'
+PREFIX = '/usr/local/openvpn'
+
+SOURCES_DIR = os.path.abspath('./sources')
+BUILD_DIR = os.path.abspath('./build')
+STAGE_DIR = os.path.abspath('./stage')
+IMAGE_DIR = os.path.abspath('./images')
+
+OPENSSL_CONFIG_PLATFORM_SCRIPT = os.path.abspath('gentoo-config-0.9.8')
+RES_PRIVACY_TXT = os.path.abspath('PRIVACY.txt')
+RES_INITIAL_SETUP = os.path.abspath('initialsetup.sh')
+RES_PQ_OPENVPN_SERVICE = os.path.abspath('pq-openvpn.service')
+
+VCVARSALL = (
+    '"C:\\Program Files (x86)\\Microsoft Visual Studio\\2017\\'
+    'Enterprise\\VC\\Auxiliary\\Build\\vcvarsall.bat"'
+)
+VERBOSE = False
+
+LINUX_UBUNTU_POST = ('''
+The staged tarball provides a readily deployable set of binaries on a Linux VM
+to quickly bring up a VPN server. It has been tested with the Ubuntu image
+currently provided by Azure. This installation may be usable as a client with
+a client configuration file instead, but this is untested, and the automatic
+service startup is configured to look for server.ovpn as a config file. To use
+the staged Linux tarball, do the following as root/using sudo in your VM:
+
+1. cd /
+2. tar xvzf <path>/pq-openvpn-linux-staged.tar.gz
+3. Create /usr/local/openvpn/etc/server.ovpn and dependent cert/key files as
+   needed.
+4. /usr/local/openvpn/sbin/initialsetup.sh
+
+To upgrade an existing installation:
+1. systemctl stop pq-openvpn
+2. cd /
+3. tar xvzf <path>/pq-openvpn-linux-staged.tar.gz
+4. systemctl start pq-openvpn
+''')
+
+
+#
+# Private Utility Methods
+# they should be shared with other python scripts to avoid
+# repeating code, errors, documentation
+#
+
+
+@contextmanager
+def chdir(path):
+    '''
+    chdir helper that returns to the cwd when the context is left.
+    avoids chdir('somewhere');do_something();chdir('..')
+    '''
+    oldcwd = os.getcwd()
+    os.chdir(path)
+    try:
+        yield
+    finally:
+        os.chdir(oldcwd)
+
+
+def die(message):
+    '''
+    Display a message and stop the build with a nonzero exit code.
+    '''
+    sys.stderr.write(message)
+    sys.stderr.write('\n')
+    sys.exit(1)
 
-VCVARSALL = '"C:\\Program Files (x86)\\Microsoft Visual Studio\\2017\\Enterprise\\VC\\Auxiliary\\Build\\vcvarsall.bat"'
 
-# Run an external command, block until it completes 
-def run_command(cmd):
-    print '***** Running command: %s' % ' '.join(map(str,cmd))
-    p = subprocess.Popen(cmd)
-    p.wait()
-
-# Clone a git repo, using the default name, in the CWD
-# If branch is specified, clone that branch
 def git_clone(repo_url, branch, local_name, commit=None):
-    r = re.compile(".*/(.*)$")
+    '''
+    Clone a git repo, using the default name, in the CWD
+    If branch is specified, clone that branch
+    '''
+    r = re.compile('.*/(.*)$')
     m = r.match(repo_url)
     repo_name = m.group(1)
-    print "Cloning %s ..." % repo_name
+    if os.path.isdir(local_name):
+        return git_pull(local_name)
```

Comment:
as long as windows commit and linux commit are the same this is ok. because the git is packed after download (git_store/git_restore) always the right commit is used (and can be verified)
changed the code to die when the directory exists with a --skip-download hint.